src/libraries/SafeCast.sol

```diff
@@ -10,31 +10,74 @@ library SafeCast {
     /// @param y The uint256 to be downcasted
     /// @return z The downcasted integer, now type uint160
     function toUint160(uint256 y) internal pure returns (uint160 z) {
-        z = uint160(y);
-        if (z != y) revert SafeCastOverflow();
+        /// @solidity memory-safe-assembly
+        assembly {
```

Comment:
couple of questions...
1. is it cheaper to remove the creation of `z` - i.e. `returns (uint160)`, and then just add `return y`?

Reply:
The following produces the same gas.
```solidity
    function toUint160(uint256 y) internal pure returns (uint160) {
        /// @solidity memory-safe-assembly
        assembly {
            if iszero(lt(y, shl(160, 1))) {
                // revert SafeCastOverflow();
                mstore(0, 0x93dafdf1)
                revert(0x1c, 0x04)
            }
        }
        return uint160(y);
    }
```

Reply:
Bumping this up since readability improved a lot.